torch/csrc/jit/codegen/cuda/lower2device.cpp

```diff
@@ -19,22 +20,163 @@
+std::unordered_map<Val*, Val*> getSimplificationMap(Fusion* fusion) {
+  std::list<std::unordered_set<IterDomain*>> disjoint_root_sets;
+  std::unordered_map<IterDomain*, std::unordered_set<IterDomain*>*>
+      id_to_disjoint_root_set;
+
+  auto map_root_ids = [&disjoint_root_sets, &id_to_disjoint_root_set](
+                          IterDomain* id0, IterDomain* id1) {
+    if (id0->isBroadcast() || id1->isBroadcast()) {
+      return;
+    }
+
+    auto disjoint_set_0_it = id_to_disjoint_root_set.find(id0);
+    auto disjoint_set_1_it = id_to_disjoint_root_set.find(id1);
+    bool set_0_found = disjoint_set_0_it != id_to_disjoint_root_set.end();
+    bool set_1_found = disjoint_set_1_it != id_to_disjoint_root_set.end();
+
+    if (set_0_found && set_1_found) {
+      if (disjoint_set_0_it->second == disjoint_set_1_it->second) {
+        return;
+      }
+      // merge second disjoint set into first
+      auto* set_0 = disjoint_set_0_it->second;
+      auto* set_1 = disjoint_set_1_it->second;
+      for (auto id : *set_1) {
+        set_0->emplace(id);
+        id_to_disjoint_root_set[id] = set_0;
+      }
+      // remove second set from disjoint_root_sets
+      disjoint_root_sets.erase(std::find(
+          disjoint_root_sets.begin(), disjoint_root_sets.end(), *set_1));
+    } else if (set_0_found || set_1_found) {
+      auto existing_set =
+          set_0_found ? disjoint_set_0_it->second : disjoint_set_1_it->second;
+      auto to_add_id = set_0_found ? id1 : id0;
+      existing_set->emplace(to_add_id);
+      id_to_disjoint_root_set[to_add_id] = existing_set;
+      // add entry into existing set
+    } else {
+      // create new set entry
+      disjoint_root_sets.push_back(std::unordered_set<IterDomain*>());
+      auto* new_set = &disjoint_root_sets.back();
+      new_set->emplace(id0);
+      new_set->emplace(id1);
+      id_to_disjoint_root_set[id0] = new_set;
+      id_to_disjoint_root_set[id1] = new_set;
+    }
+  };
+
+  auto fusion_vals = fusion->usedMathVals();
+  for (auto producer_tv : ir_utils::filterByType<TensorView>(fusion_vals)) {
+    auto consumer_tvs = scheduler_utils::consumerTvsOf({producer_tv});
+    for (auto consumer_tv : consumer_tvs) {
+      auto pairwise_map = PairwiseRootDomainMap(producer_tv, consumer_tv);
+      auto c2p_root_map = pairwise_map.mapConsumerToProducer(
+          consumer_tv->domain(), producer_tv->domain());
+      for (auto entry : c2p_root_map) {
+        auto c_id = entry.first;
+        auto p_id = entry.second;
+        map_root_ids(p_id, c_id);
+      }
+    }
+  }
+
+  // Map each set to an input ID (if it exists) that has the smallest ->name()
+  // entry value
+  std::unordered_map<std::unordered_set<IterDomain*>*, IterDomain*>
+      set_to_input_id;
 
+  for (auto input_tv : ir_utils::filterByType<TensorView>(fusion->inputs())) {
+    for (auto id :
+         TensorDomain::noReductions(input_tv->getMaybeRFactorDomain())) {
```

Comment:
For rFactor tensors, would it also make sense to look at non-rFactor root domains as well? IIUC, non-rFactor root domains could have replacements.

Reply:
We're looping over input tensors of the Fusion, so I actually don't think we need RFactor, but put it in because I don't think it hurts. We're looking forward from the inputs only so I thought RFactor would be fine since that's the only real root link to consumers.

In practice, this would only make a difference if we did rfactor before segmentation, I don't know if there's a practical reason to do that.

Reply:
Need to add comment about why we only look at inputs.